src/futures/digital.rs

```diff
@@ -0,0 +1,50 @@
+//! Asynchronous digital I/O
+
+use core::future::Future;
+
+use crate::blocking::digital::InputPin;
+
+/// Asynchronously wait for a pin to become high or low.
+///
+/// # Examples
+/// ```rust
+/// # use embedded_hal::futures::digital::AsyncInputPin;
+/// /// Asynchronously wait until the `ready_pin` becomes high.
+/// async fn wait_until_ready<P>(ready_pin: &P)
+/// where
+///     P: AsyncInputPin,
+/// {
+///     ready_pin.until_high().await
+/// }
+/// ```
+///
+/// ```rust,ignore
+/// # use embedded_hal::futures::digital::AsyncInputPin;
+/// # use embedded_hal::futures::delay::Delay;
+/// # use core::time::Duration;
+/// /// Wait until the `ready_pin` is high or timeout after 1 millisecond.
+/// /// Returns true is the pin became high or false if it timed-out.
+/// async fn wait_until_ready_or_timeout<P, D>(ready_pin: &P, delay: &mut D) -> bool
+/// where
+///     P: AsyncInputPin,
+///     D: Delay,
+/// {
+///     futures::select_biased! {
+///         _ => ready_pin.until_high() => true,
+///         _ => delay.delay(Duration::from_millis(1)) => false,
+///     }
+/// }
+/// ```
+pub trait AsyncInputPin: InputPin {
+    /// The future returned by the `until_high` function.
+    type UntilHighFuture<'a>: Future<Output=()> + 'a;
+
+    /// The future returned by the `until_low` function.
+    type UntilLowFuture<'a>: Future<Output=()> + 'a;
+
+    /// Returns a future that resolves when this pin becomes high.
+    fn until_high<'a>(&self) -> Self::UntilHighFuture<'a>;
```

Comment:
Requiring `&self` instead of `&mut self` means multiple tasks can wait on the same pin concurrently. This will require implementations to support storing an arbitrary amount of wakers, which is generally not feasible.

Reply:
Good point, hadn't thought of that.

What do you think of the naming? Is there a better name than `until_high|low`?

Reply:
just commented about the names in a new thread